<a href="https://colab.research.google.com/github/ai7dnn/2026-1-BDA/blob/main/code/04w_array_stack_split_B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

a = np.arange(1, 13).reshape(3, 4)
c = np.arange(1, 6, 2)
a, c

(array([[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]]),
 array([1, 3, 5]))

In [2]:
c[:, None]

array([[1],
       [3],
       [5]])

In [3]:
c[:, np.newaxis]

array([[1],
       [3],
       [5]])

In [5]:
cc = np.expand_dims(c, axis=1)
cc

array([[1],
       [3],
       [5]])

In [7]:
np.hstack( (a, cc) )

array([[ 1,  2,  3,  4,  1],
       [ 5,  6,  7,  8,  3],
       [ 9, 10, 11, 12,  5]])

In [8]:
np.hstack( (a, c) )

ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)

In [9]:
np.column_stack((a, c))

array([[ 1,  2,  3,  4,  1],
       [ 5,  6,  7,  8,  3],
       [ 9, 10, 11, 12,  5]])

In [10]:
np.column_stack((a, cc))

array([[ 1,  2,  3,  4,  1],
       [ 5,  6,  7,  8,  3],
       [ 9, 10, 11, 12,  5]])

In [11]:
a = np.arange(1, 4)
b = np.arange(4, 7)
a, b

(array([1, 2, 3]), array([4, 5, 6]))

In [15]:
np.c_[a, b], np.c_[ (a, b) ]

(array([[1, 4],
        [2, 5],
        [3, 6]]),
 array([[1, 4],
        [2, 5],
        [3, 6]]))

In [14]:
np.c_[a, b], np.column_stack((a, b))

(array([[1, 4],
        [2, 5],
        [3, 6]]),
 array([[1, 4],
        [2, 5],
        [3, 6]]))

In [18]:
x = np.arange(1, 4).reshape(1, 3)
y = np.arange(4, 7).reshape(1, 3)
x, y

(array([[1, 2, 3]]), array([[4, 5, 6]]))

In [19]:
np.c_[x, y]

array([[1, 2, 3, 4, 5, 6]])

In [20]:
np.column_stack((x, y))

array([[1, 2, 3, 4, 5, 6]])

In [22]:
np.c_[x, 10, 20, y]

array([[ 1,  2,  3, 10, 20,  4,  5,  6]])

In [23]:
np.column_stack((x, 10, 20, y))

array([[ 1,  2,  3, 10, 20,  4,  5,  6]])

### numpy 함수 vstack hstack row_stack column_stack 비교 설명

| 항목 | `np.vstack` | `np.hstack` | `np.row_stack` | `np.column_stack` |
|---|---|---|---|---|
| **별칭 여부** | — | — | `vstack`의 alias | — |
| **연결 축** | axis=0 (수직) | axis=1 (수평) | axis=0 (수직) | axis=1 (수평) |
| **내부 구현** | `concatenate(axis=0)` + `atleast_2d` | 1D: `concatenate(axis=0)` / 2D+: `concatenate(axis=1)` | `vstack`과 동일 | 1D를 `(N,1)`로 변환 후 `concatenate(axis=1)` |
| **shape 요구 조건** | 열 수 일치 | 행 수 일치 | 열 수 일치 | 행 수 일치 |
| **Deprecated** | — | — | NumPy 2.0+ | — |

In [172]:
a = np.arange(1, 4)
b = np.arange(4, 7)
a, b

(array([1, 2, 3]), array([4, 5, 6]))

In [173]:
x = np.arange(1, 7).reshape(2, 3)
y = np.arange(7, 13).reshape(2, 3)
x, y

(array([[1, 2, 3],
        [4, 5, 6]]),
 array([[ 7,  8,  9],
        [10, 11, 12]]))

In [174]:
np.vstack((a, b)) # 1D --> 2D

array([[1, 2, 3],
       [4, 5, 6]])

In [175]:
np.hstack((a, b)) # 1D --> 1D

array([1, 2, 3, 4, 5, 6])

In [176]:
np.vstack((x, y)) # 2D --> 2D

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [179]:
np.hstack((x, y)) # 2D --> 2D

array([[ 1,  2,  3,  7,  8,  9],
       [ 4,  5,  6, 10, 11, 12]])

In [181]:
a, b

(array([1, 2, 3]), array([4, 5, 6]))

In [180]:
np.column_stack((a, b)) # 1D --> 2D

array([[1, 4],
       [2, 5],
       [3, 6]])

In [182]:
np.column_stack((a, b)), np.c_[a, b] # 1D --> 2D

(array([[1, 4],
        [2, 5],
        [3, 6]]),
 array([[1, 4],
        [2, 5],
        [3, 6]]))

In [184]:
np.column_stack((x, y)), np.c_[x, y] # 2D --> 2D

(array([[ 1,  2,  3,  7,  8,  9],
        [ 4,  5,  6, 10, 11, 12]]),
 array([[ 1,  2,  3,  7,  8,  9],
        [ 4,  5,  6, 10, 11, 12]]))

In [186]:
m = np.arange(24).reshape(2, 3, 4)
m

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [189]:
n = np.arange(96).reshape(4, 4, 6)
n

array([[[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]],

       [[24, 25, 26, 27, 28, 29],
        [30, 31, 32, 33, 34, 35],
        [36, 37, 38, 39, 40, 41],
        [42, 43, 44, 45, 46, 47]],

       [[48, 49, 50, 51, 52, 53],
        [54, 55, 56, 57, 58, 59],
        [60, 61, 62, 63, 64, 65],
        [66, 67, 68, 69, 70, 71]],

       [[72, 73, 74, 75, 76, 77],
        [78, 79, 80, 81, 82, 83],
        [84, 85, 86, 87, 88, 89],
        [90, 91, 92, 93, 94, 95]]])

In [190]:
np.split(n, 2)

[array([[[ 0,  1,  2,  3,  4,  5],
         [ 6,  7,  8,  9, 10, 11],
         [12, 13, 14, 15, 16, 17],
         [18, 19, 20, 21, 22, 23]],
 
        [[24, 25, 26, 27, 28, 29],
         [30, 31, 32, 33, 34, 35],
         [36, 37, 38, 39, 40, 41],
         [42, 43, 44, 45, 46, 47]]]),
 array([[[48, 49, 50, 51, 52, 53],
         [54, 55, 56, 57, 58, 59],
         [60, 61, 62, 63, 64, 65],
         [66, 67, 68, 69, 70, 71]],
 
        [[72, 73, 74, 75, 76, 77],
         [78, 79, 80, 81, 82, 83],
         [84, 85, 86, 87, 88, 89],
         [90, 91, 92, 93, 94, 95]]])]

In [194]:
np.split(n, 2)[1].shape

(2, 4, 6)

In [195]:
np.split(n, 2, axis=1)

[array([[[ 0,  1,  2,  3,  4,  5],
         [ 6,  7,  8,  9, 10, 11]],
 
        [[24, 25, 26, 27, 28, 29],
         [30, 31, 32, 33, 34, 35]],
 
        [[48, 49, 50, 51, 52, 53],
         [54, 55, 56, 57, 58, 59]],
 
        [[72, 73, 74, 75, 76, 77],
         [78, 79, 80, 81, 82, 83]]]),
 array([[[12, 13, 14, 15, 16, 17],
         [18, 19, 20, 21, 22, 23]],
 
        [[36, 37, 38, 39, 40, 41],
         [42, 43, 44, 45, 46, 47]],
 
        [[60, 61, 62, 63, 64, 65],
         [66, 67, 68, 69, 70, 71]],
 
        [[84, 85, 86, 87, 88, 89],
         [90, 91, 92, 93, 94, 95]]])]

In [196]:
np.split(n, 2, axis=2)

[array([[[ 0,  1,  2],
         [ 6,  7,  8],
         [12, 13, 14],
         [18, 19, 20]],
 
        [[24, 25, 26],
         [30, 31, 32],
         [36, 37, 38],
         [42, 43, 44]],
 
        [[48, 49, 50],
         [54, 55, 56],
         [60, 61, 62],
         [66, 67, 68]],
 
        [[72, 73, 74],
         [78, 79, 80],
         [84, 85, 86],
         [90, 91, 92]]]),
 array([[[ 3,  4,  5],
         [ 9, 10, 11],
         [15, 16, 17],
         [21, 22, 23]],
 
        [[27, 28, 29],
         [33, 34, 35],
         [39, 40, 41],
         [45, 46, 47]],
 
        [[51, 52, 53],
         [57, 58, 59],
         [63, 64, 65],
         [69, 70, 71]],
 
        [[75, 76, 77],
         [81, 82, 83],
         [87, 88, 89],
         [93, 94, 95]]])]

### 난수 기초와 생성

In [55]:
import numpy as np

rg = np.random.default_rng(12)
rg.random() # [0, 1) 실수 난수

0.2508244581084461

In [70]:
rg.random(4)

array([0.43307411, 0.40593661, 0.8860097 , 0.31372551])

In [71]:
rg.random( (3, 4, 2) )

array([[[0.54643986, 0.56456315],
        [0.04331116, 0.09893088],
        [0.2360827 , 0.00272549],
        [0.68328311, 0.57861684]],

       [[0.34200102, 0.41912567],
        [0.04097307, 0.40095732],
        [0.84128334, 0.0378487 ],
        [0.65314676, 0.69133897]],

       [[0.21925692, 0.2954974 ],
        [0.17960469, 0.56591141],
        [0.36929492, 0.13015358],
        [0.28901536, 0.24438015]]])

In [74]:
import numpy as np

rg = np.random.default_rng(12)
rg.integers(1, 7, (3, 4)) # [1, 7) 정수 난수

array([[4, 2, 6, 6],
       [1, 2, 2, 2],
       [4, 3, 3, 2]])

In [88]:
rg.integers(low=1, high=10, size=(2, 8)) # [1, 10) 정수 난수

array([[1, 1, 2, 7, 6, 7, 3, 7],
       [9, 5, 2, 5, 1, 7, 5, 3]])

In [94]:
import numpy as np

rg = np.random.default_rng(12)
rg.standard_normal(10) # 표준 정규 분포 실수 난수

array([-0.00682678,  1.04614329,  0.74158842,  0.72395654,  1.61877622,
       -1.20555814, -0.62695547, -1.32066321, -0.10775251,  0.99876366])

In [103]:
rg.standard_normal( (3, 4) )

array([[ 0.04838089,  1.02063667, -0.02489578, -0.25996794],
       [ 0.53864522,  2.07661706, -0.34274768,  0.51286725],
       [-1.57730495,  1.34549387,  0.65938351,  1.33279967]])

In [122]:
rg.standard_normal( (10, 5) )

array([[-2.74374861,  1.01292039,  0.50721526,  0.29578685,  0.86171368],
       [ 0.32965999, -0.95408306, -0.67916811, -0.16717385,  0.91878933],
       [ 0.03302551, -0.66878885,  0.3645885 , -1.23718444,  2.09023409],
       [ 0.30413237, -0.99181179,  0.16536093,  1.08116344,  0.54186694],
       [ 2.04640067, -1.43037405,  1.59821622, -0.03015015,  0.8851731 ],
       [ 1.72398429, -1.20721724, -1.42387392, -0.39219578, -1.67239514],
       [-2.02805951,  0.08281067,  1.17027128, -0.48033556,  0.50636717],
       [ 3.70039125,  0.33989688, -0.30407771, -0.15893203,  0.43833025],
       [-0.28045033,  0.70265275,  0.22796743,  0.93298251,  0.16055557],
       [-0.07021734, -0.92409945, -1.3797935 , -0.27538646, -0.37907176]])

### np.random.random()	np.random.default_rng().random() 비교

| 항목    | np.random.random() | default_rng().random() |
| ----- | ------------------ | ---------------------- |
| 방식    | 전역 상태 사용           | 객체 기반                  |
| 알고리즘  | Mersenne Twister   | PCG64                  |
| 재현성   | 취약 (전역 영향 받음)      | 매우 안정적                 |
| 병렬 처리 | 부적합                | 적합                     |
| 권장 여부 | ❌ (구식)             | ✅ (권장)                 |


np.random.random() np.random.default_rng().random() 비교

| 항목 | `np.random.random()` (Legacy / RandomState) | `default_rng().random()` (New / Generator, NumPy 1.17+) |
|---|---|---|
| **내부 알고리즘** | Mersenne Twister (MT19937) | PCG64 (기본값); SFC64, MT19937 등 선택 가능 |
| **상태 크기** | 624 × 32-bit words (~2.5 KB) | PCG64: 128-bit state (16 B) |
| **주기 (period)** | 2^19937 − 1 | PCG64: 2^128 |
| **BigCrush 통과** | 일부 실패 (LinearComp, MatrixRank 등) | 통과 |
| **고차원 균일성** | MT는 623차원에서 균일, 그 이상 편향 가능 | PCG64: 고차원에서도 우수 |
| **시드 설정** | `np.random.seed(42)` — 전역 상태 변경 (side-effect) | `rng = np.random.default_rng(42)` — 로컬 객체, 독립적 |
| **전역 상태 오염** | 있음 — 다른 코드와 간섭 위험 | 없음 — 인스턴스 독립 |
| **버전 간 재현성** | NumPy 버전 간 보장 안 됨 | 동일 BitGenerator 사용 시 보장 |
| **독립 스트림 생성** | 불편 — 수동 시드 관리 필요 | 공식 지원 (`rng.spawn(n)` 또는 `SeedSequence`) |
| **속도** | 보통 | PCG64 > MT19937 일반적; SFC64는 더 빠름 |
| **추가 분포 API** | 기본 분포 제공 | `rng.integers()`, `rng.standard_normal()` 등 확장 API |
| **BitGenerator 교체** | 불가 | 가능 — `np.random.Generator(MT19937(42))` |
| **신규 코드 권장** | 비권장 | 권장 (NEP 19) |
| **레거시 호환** | 적합 | 일부 함수명 상이 (하위 호환 주의) |
| **ML/시뮬레이션 연구** | 재현성 관리 복잡 | 권장 — 실험 단위 RNG 분리 용이 |

In [148]:
import numpy as np

# np.random.seed(12)
np.random.random() # [0, 1) 실수 난수

0.9007148541170122

In [149]:
from numpy import random

# np.random.seed(12)
random.random() # [0, 1) 실수 난수

0.03342142762634459

In [151]:
import numpy as np

np.random.randint(1, 10, 5)

array([6, 9, 3, 4, 5])

In [161]:
np.random.rand(3, 4)

array([[0.40284721, 0.88517769, 0.20215434, 0.41901483],
       [0.64795917, 0.51164295, 0.33383249, 0.77847343],
       [0.13543513, 0.19171077, 0.40258915, 0.78684447]])

In [170]:
np.set_printoptions(precision=3)

In [171]:
np.random.rand(5, 3)

array([[0.999, 0.671, 0.79 ],
       [0.982, 0.142, 0.026],
       [0.746, 0.617, 0.099],
       [0.522, 0.507, 0.286],
       [0.248, 0.18 , 0.584]])